test/sqllogictest/transactions.slt

```diff
@@ -799,3 +825,108 @@ simple
 SHOW SOURCES
 ----
 COMPLETE 0
+
+# Test that the cluster cannot change mid-transaction.
+
+statement ok
+DROP TABLE t CASCADE
+
+statement ok
+CREATE TABLE t (a INT)
+
+statement ok
+CREATE CLUSTER c REPLICAS (r1 (SIZE '1'))
+
+statement ok
+BEGIN
+
+statement ok
+SELECT * FROM t
+
+statement error SET cluster cannot be called in an active transaction
+SET CLUSTER TO c
+
+statement ok
+ROLLBACK
+
+# Test that the cluster is selected at the start of a transaction and doesn't change.
+
+## Auto-routing selects mz_introspection at the start of transaction.
+
+statement ok
+BEGIN
+
+statement ok
+SHOW VIEWS
+
+query T multiline
+EXPLAIN SELECT *, generate_series(1, 100) FROM mz_views
+----
+Explained Query:
+  CrossJoin type=differential
+    ArrangeBy keys=[[]]
+      ReadIndex on=mz_views mz_show_views_ind=[*** full scan ***]
+    ArrangeBy keys=[[]]
+      Constant
+        total_rows (diffs absed): 100
+        first_rows:
+          - (1)
+          - (2)
+          - (3)
+          - (4)
+          - (5)
+          - (6)
+          - (7)
+          - (8)
+          - (9)
+          - (10)
+          - (11)
+          - (12)
+          - (13)
+          - (14)
+          - (15)
+          - (16)
+          - (17)
+          - (18)
+          - (19)
+          - (20)
+
+Used Indexes:
+  - mz_internal.mz_show_views_ind (*** full scan ***)
+
+EOF
+
+statement ok
+COMMIT
+
+## Auto-routing doesn't select mz_introspection in the middle of a transaction.
+
+# Since mz_views uses custom types, the postgres client will look it up in the catalog on
+# first use. If the first use happens to be in a transaction, then we can get unexpected time
+# domain errors. This is an annoying hack to load the information in the postgres client before
+# we start any transactions.
+statement ok
+SELECT * FROM mz_views LIMIT 0
+----
```

Comment:
This is still required with this change due to https://github.com/MaterializeInc/materialize/issues/21815#issuecomment-1729712813, which is very unfortunate (CC @mjibson).